src/Facet/Env.hs

```diff
@@ -0,0 +1,40 @@
+module Facet.Env
+( Env(..)
+, empty
+, (|>)
+, lookup
+, index
+, level
+) where
+
+import Control.Applicative ((<|>))
+import Control.Monad (guard)
+import Data.Maybe (fromMaybe)
+import Facet.Core.Pattern
+import Facet.Name
+import Facet.Snoc
+import Facet.Syntax
+import GHC.Stack
+import Prelude hiding (lookup)
+
+newtype Env v = Env { bindings :: Snoc (Pattern (Name :=: v)) }
+  deriving (Foldable, Functor, Monoid, Semigroup, Traversable)
```

Comment:
I’ve been meaning to define this for a while since it can be used for normalization as well as for printing and has a bunch of conveniences relative to bare-bones `Snoc`.